Clean up isinstance flags

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33265 Clean up isinstance flags**
* #32889 Move special ops into interpreter
* #32804 Remove prim::Constant op
* #32791 interpreter handling for varargs to remove need for looking at Node

This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

Differential Revision: [D19883933](https://our.internmc.facebook.com/intern/diff/D19883933)

zdevito committed:
Clean up isinstance flags

This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1852,6 +1857,55 @@ static ScalarTypeTypePtr get();
+};
+
+// the common supertype of all tuples,
+// Tuple[T...] <: AnyList for all T
```
typo s/AnyList/AnyTuple/

bhosmer commented on aten/src/ATen/core/type.cpp:
```diff
@@ -560,8 +570,12 @@ TupleType::TupleType(
     return true;
+  }
   auto rhs = rhs_->cast<TupleType>();
   if (!rhs)
```
If this test is isomorphic to the new one, we should probably standardize on one and kill the other (or maybe leave it as an assert, I guess)

...and if there's a motivation for leaving them both in, maybe worth noting it in a comment, as it's not super obvious.

zdevito replied:
AnyTupleType doesn't cast to TupleType. It would be hard to make that work since TupleType is also expected to list its members. I think this is what we want for now because use of AnyTupleType is constrained only to isinstance checks.

bhosmer reviewed (approved):
SWEET

bhosmer commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -12,10 +12,10 @@ namespace mobile {
   TORCH_CHECK(isOpSupportedInMobile(op), toString(op),
               " is not supported in mobile module.");
-  code_->instructions_.emplace_back(op, X, N, flags);
+  code_->instructions_.emplace_back(op, X, N);
```
Just noting that the original had an explicit `0` as the fourth argument here

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

Differential Revision: [D19883933](https://our.internmc.facebook.com/intern/diff/D19883933)

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

Differential Revision: [D19883933](https://our.internmc.facebook.com/intern/diff/D19883933)

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

Differential Revision: [D19883933](https://our.internmc.facebook.com/intern/diff/D19883933)

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

Differential Revision: [D19883933](https://our.internmc.facebook.com/intern/diff/D19883933)

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

Differential Revision: [D19883933](https://our.internmc.facebook.com/intern/diff/D19883933)

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

Differential Revision: [D19883933](https://our.internmc.facebook.com/intern/diff/D19883933)

[ghstack-poisoned]

zdevito committed:
Update on "Clean up isinstance flags"


This removes the need for isinstance to keep trace of list and tuple
separately by introducing AnyListType and AnyTupleType into the JIT
type system to be the common supertype of any lists or tuples.

This allows us to remove the weird flags from the interpreter for
the isinstance operator.

Differential Revision: [D19883933](https://our.internmc.facebook.com/intern/diff/D19883933)

[ghstack-poisoned]
